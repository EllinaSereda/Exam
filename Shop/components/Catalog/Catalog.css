.Catalog img{
    width:80%;
    
}
.Catalog .Product{
    width:100%;
    padding:50px 0;
    text-align: center;
}
.Catalog .Product:hover{
    box-shadow: 0 0 20px rgba(0,0,0,0.3);
}
.Catalog:after{
    clear:both;
    display: block;
    content:"";
}
.Catalog {
    display:grid;
    width:75%;
    grid-template-columns: repeat(3, 1fr);
}


.Catalog .Product h3{
    font-size: 22px;
    color:rgb(84, 84, 84);
    color:rgba(0, 0, 0, 0.3);
    margin-top:25px;
    margin-bottom:15px;
}
.Catalog .Product p{
    font-size: 20px;
    padding:0 15px;
    margin-bottom:30px;
}
.Catalog .Product .Volume{
    font-size: 16px;
    padding:0 15px;
    margin-bottom:30px;
}
.Catalog  .Product .Volume span, .Catalog  .Product .Volume .Var{
    color:rgb(84, 84, 84);
    color:rgba(0, 0, 0, 0.3);
}
.Catalog .Product .Price{
    font-size: 20px;
    padding:0 15px;
    margin-bottom:30px;
    font-weight:bold;
}
.Catalog  .Product .hidden{
    display: none;
}
.Catalog  .Product:hover .hidden{
    display:block;
}
.Catalog  .Product:hover .hidden:hover{
    background-color:rgba(156, 60, 118, 1);
}
.Catalog  .Product .hidden{
    font-family: 'Ubuntu', sans-serif;
    background-color:rgba(156, 60, 118, 0.7);
    color:white;
    border-radius:10px;
    font-size:18px;
    padding:10px 20px;
    margin:10px auto;
}
@media screen and (orientation:portrait) and  (max-device-width:736pt) {
    .Catalog  {
        width:100%;
    }


}
@media screen and  (max-width:650pt) {
    .Catalog {
        width:100%;
    }
}