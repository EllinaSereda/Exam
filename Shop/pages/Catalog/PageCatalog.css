
.PageCatalog:after{
    clear:both;
    display: block;
    content:"";
}

.PageCatalog .Show{
    float:right;
    margin:20px;
    margin-right:40px;

}
.PageCatalog .Show span, .PageCatalog .Pages span{
    padding:0 5px;
    font-size:18px;
    color:#631B42;
}
.PageCatalog .Show span.active, .PageCatalog .Pages span.active{
    font-weight:bold;
    border-bottom: 1px dashed #631B42;

}
.PageCatalog select{
    float:right;
    margin:20px;
    border:none;
    background-color:white;
    padding: 10px;
    font-size:18px;
    outline:none;
    box-shadow: none;
}
.PageCatalog .Pages{
    font-size:20px;
    width:75%;
    float:right;
    text-align: center;
    margin:30px 0;
}

.PageCatalog .Pages span{
    font-size:20px;
    
}

@media screen and (orientation:portrait) and  (max-device-width:736pt) {

    .PageCatalog select{
        float:none;
        margin:10px;
        padding: 10px;
        font-size:30px;
    }
    .PageCatalog .Show span, .PageCatalog .Pages span{
        padding:0 5px;
        font-size:35px;
    }
    .PageCatalog .Pages{
        width:100%;
        text-align: center;
        margin:30px 0;
    }


}
@media screen and  (max-width:650pt) {
  
    .PageCatalog .Pages{
        width:100%;
        text-align: center;
    }
}