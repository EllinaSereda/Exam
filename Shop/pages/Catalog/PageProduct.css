.PageProduct .Product{
    padding:15px;
    text-align: center;
}
.PageProduct:after{
    clear:both;
    display: block;
    content:"";
}
.PageProduct .describe{
    width:80%;
    text-align:left;
    margin:80px auto;
    
}

.PageProduct .describe .title{
    font-size:24px;
    font-weight: bold;
    margin-right:30px;
    padding:20px 5px;
    color:#616668;
    
}
.PageProduct .describe .active{
    color:black;
    border-bottom:4px solid #9C3C76;
}

.PageProduct .describe .content{
    font-size:18px;
}
.PageProduct .describe .Part{
    font-size:18px;
    padding:20px 0;
    border-bottom:1px solid rgba(0, 0, 0, 0.3);;
}
.PageProduct .describe p{
    margin:10px 0;
}
.PageProduct .describe p .bold{
    font-weight:bold;
    font-size:20px;
}


.PageProduct .ProdBrand>div {
    display:grid;
    width:80%;
    margin:auto;
    grid-template-columns: repeat(3, 1fr);
}
.PageProduct .ProdBrand h6 {
    font-size:24px;
    
}
.PageProduct .ProdBrand .CatalogProduct img{
    width:80%;
    
}
.PageProduct .ProdBrand .CatalogProduct{
    width:100%;
    padding:50px 0;
    text-align: center;
}
.PageProduct .ProdBrand .CatalogProduct:hover{
    box-shadow: 0 0 20px rgba(0,0,0,0.3);
}

.PageProduct .ProdBrand .CatalogProduct h3{
    font-size: 22px;
    color:rgb(84, 84, 84);
    color:rgba(0, 0, 0, 0.3);
    margin-top:25px;
    margin-bottom:15px;
}
.PageProduct .ProdBrand .CatalogProduct p{
    font-size: 20px;
    padding:0 15px;
    margin-bottom:30px;
}
.PageProduct .ProdBrand .CatalogProduct .Volume{
    font-size: 16px;
    padding:0 15px;
    margin-bottom:30px;
}
.PageProduct .ProdBrand .CatalogProduct .Volume span, .PageProduct .CatalogProduct .Volume .Var{
    color:rgb(84, 84, 84);
    color:rgba(0, 0, 0, 0.3);
}
.PageProduct .ProdBrand .CatalogProduct .Price{
    font-size: 20px;
    padding:0 15px;
    margin-bottom:30px;
    font-weight:bold;
}
.PageProduct .ProdBrand .CatalogProduct .hidden{
    display: none;
}
.PageProduct .ProdBrand .CatalogProduct:hover .hidden{
    display:block;
}
.PageProduct .ProdBrand .CatalogProduct:hover .hidden:hover{
    background-color:rgba(156, 60, 118, 1);
}
.PageProduct .ProdBrand .CatalogProduct .hidden{
    font-family: 'Ubuntu', sans-serif;
    background-color:rgba(156, 60, 118, 0.7);
    color:white;
    border-radius:10px;
    font-size:18px;
    padding:10px 20px;
    margin:10px auto;
}
