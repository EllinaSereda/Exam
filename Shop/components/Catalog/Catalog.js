import React from 'react';
import PropTypes from 'prop-types';
import './Catalog.css';
import { NavLink } from 'react-router-dom';
class Catalog extends React.PureComponent {

  static propTypes = {

    
  }



  //Устанавка  первой страницы после применения фильтров


  // Создаем массив брендов всех товаров


  render() {
    return <div className="Catalog">
    { this.props.products.map(v=> <div key={v.code} className='Product'>
        <NavLink key={v.code} to={"/product/"+v.code} exact className="PageLink" activeClassName="ActivePageLink"><img alt={0} key={0}  src={v.img[0]}/>
        <h3>{v.brand}</h3>
        <p>{v.name}</p>
        <div className="Volume">Объем, мл: {v.stock.sort((a,b)=>a.vol-b.vol).map(v=><span key={v.vol}>{v.vol} </span>)}
        <div className="Var">{v.stock.length} варианта</div></div>
        <div className="Price">{Math.min.apply(null,v.stock.map(v=>v.price))} - {Math.max.apply(null,v.stock.map(v=>v.price))} руб.</div>
        <input type="button" value="Выбрать" className="hidden"/>
        </NavLink>
    </div>)
    }
    </div>
   
    

  }

}




export default Catalog;
