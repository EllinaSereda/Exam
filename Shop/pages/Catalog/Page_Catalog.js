import React from 'react';
import PropTypes from 'prop-types';
import './PageCatalog.css';
import Form from'../../components/Form';
import Catalog from '../../components/Catalog/Catalog';
import {sort} from '../../modules/sort';
import {min} from '../../modules/min';
import {minvol} from '../../modules/minvol';
import {maxvol} from '../../modules/maxvol';
import {max} from '../../modules/max';
import {brand} from '../../modules/brand';
import {sex} from '../../modules/sex';
import {connect} from 'react-redux';
import {voteEvents} from '../../events';
import { NavLink } from 'react-router-dom';
class Page_Catalog extends React.PureComponent {

  static propTypes = {
    products:PropTypes.shape({
        products:PropTypes.array,
        search:PropTypes.array,
      }),
    savedFilter:PropTypes.array,
  };
  state={
    info:null,
    bigImg:0,
    prodNumber:3,
    prodTotal:null,
    startElem:0,
    sort:0,
    upd:0,
    updP:0,
  }

  componentDidMount = () => {
    voteEvents.addListener('EFilterSaved',this.setStartPage);  //Сохранение формы
  };

  componentWillUnmount = () => {
    voteEvents.removeListener('EFilterSaved',this.setStartPage);
  };


  //Устанавка  первой страницы после применения фильтров

  setStartPage=()=>{
    this.props.history.push('1');
  }
  

  setSort=(EO)=>{
    this.setState({sort:Number(EO.target.value)});
    this.setStartPage();
  }

  // Создаем массив брендов всех товаров

  brands=()=>{
    let brands=[];
    this.props.products.products.map(v=>{
      if (!brands.some(x=>x==v.brand)){
        brands.push(v.brand)
      }
    })
    return brands.sort((a,b)=>a-b);
  }

  filterProducts=()=>{
    
    let typeP=this.props.match.params.s;
    console.log(this.props.match.params);
    let result;
    typeP?
    result=this.props.products.search:
    result=this.props.products.products;
    result=brand(result, this.props.savedFilter.savedBrands);
    result=min(result, this.props.savedFilter.sminprice);
    result=max(result, this.props.savedFilter.smaxprice);
    result=minvol(result, this.props.savedFilter.sminvolume);
    result=maxvol(result, this.props.savedFilter.smaxvolume);
    result=sex(result, this.props.savedFilter.ssex);
    return result;
  }

  sortProducts=(prod)=>{
    return sort(prod,this.state.sort);
  }
  change=(EO)=>{
    this.setState({upd:EO.target.textContent});
  }
  changeP=(EO)=>{
 
    this.setState({updP:EO.target.textContent});

  }
  




  render() {

    console.log('Page_Catalog render');
    let code;
    let products;
    products=this.filterProducts();
    products=this.sortProducts(products);
  //фильтрация по отмеченным брендам
    let sortVariants=<select defaultValue="null" onChange={this.setSort}>
      <option value="0">По популярности</option>
      <option value="1">Сортировка A-Я</option>
      <option value="2">Сортировка Я-А</option>
      <option value="3">Начинать с дешевых</option>
      <option value="4">Начинать с дорогих</option>
      <option value="5">Сортировка бренды А-Я</option>
      <option value="6">Сортировка бренды Я-А</option>
    </select>
    
    //Номера страниц
    let pages=[];  
    if (products.length/Number(this.props.match.params.amount)!=1){ 
        for(let i=1;i<=Math.ceil(products.length/Number(this.props.match.params.amount));i++){
        pages[i]=<NavLink key={i} to={"/catalog/"+this.props.match.params.amount+"/"+i} exact className="PageLink" activeClassName="ActivePageLink"><span onClick={this.changeP} className={Number(this.props.match.params.page)==i?"active":null}>{i}</span></NavLink>
        }
    }  
    
    code=products.slice(Number(this.props.match.params.amount)*(Number(this.props.match.params.page)-1),Number(this.props.match.params.amount)*(Number(this.props.match.params.page)-1) + Number(this.props.match.params.amount));
  
    console.log(code);
    return  <div className='PageCatalog' >
    <Form brands={this.brands()}/>
    <div className="Show">
    <NavLink key={4} to={"/catalog/"+4+"/"+1} exact className="PageLink" activeClassName="ActivePageLink"><span onClick={this.change} className={Number(this.props.match.params.amount)==4?"active":null}>4</span></NavLink>
    <NavLink key={5} to={"/catalog/"+5+"/"+1} exact className="PageLink" activeClassName="ActivePageLink"><span onClick={this.change} className={Number(this.props.match.params.amount)==5?"active":null}>5</span></NavLink> 
    <NavLink key={10} to={"/catalog/"+10+"/"+1} exact className="PageLink" activeClassName="ActivePageLink"><span onClick={this.change} className={Number(this.props.match.params.amount)==10?"active":null}>10</span></NavLink> 
    <NavLink key="all" to={"/catalog/"+this.props.products.products.length+"/"+1} exact className="PageLink" activeClassName="ActivePageLink"><span onClick={this.change} className={Number(this.props.match.params.amount)==this.props.products.products.length?"active":null}>all</span></NavLink> 
    </div>
    {sortVariants}
    <Catalog  products={code}/>
    <div className='Pages'>
    {Number(this.props.match.params.page)>1?<NavLink key={"prev"} to={"/catalog/"+this.props.match.params.amount+"/"+(Number(this.props.match.params.page)-1)} exact className="PageLink" activeClassName="ActivePageLink"><span onClick={this.changeP}>Предыдущая</span></NavLink>:null}
     {pages}
   {Number(this.props.match.params.page)!=(Math.ceil(products.length/Number(this.props.match.params.amount)))?<NavLink key={"nexr"} to={"/catalog/"+this.props.match.params.amount+"/"+(Number(this.props.match.params.page)+1)} exact className="PageLink" activeClassName="ActivePageLink"><span onClick={this.changeP}>Следующая</span></NavLink>:null}
    </div>
    </div>;
   
    

  }

}


const mapStateToProps = function (state) {
  return {
    savedFilter: state.formSort,  //Бренды выбранные в формк
    products: state.products,   //Товары
  };
};

export default connect(mapStateToProps)(Page_Catalog);
