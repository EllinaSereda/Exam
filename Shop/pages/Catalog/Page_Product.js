import React from 'react';
import PropTypes from 'prop-types';
import {connect} from 'react-redux';
import firebase from 'firebase';
import { NavLink } from 'react-router-dom';
import {voteEvents} from '../../events';
import {product_addtobusket} from '../../redux/UserAC';
import './PageProduct.css';
import Product from '../../components/Product/Product'
class Page_Product extends React.PureComponent {

  static propTypes = {
    brands:PropTypes.shape(
      {
        brands:PropTypes.array,
        brandsMain:PropTypes.array,
      }),
    products:PropTypes.shape({
        products:PropTypes.array,
        search:PropTypes.array,
      }),
    user:PropTypes.shape({
        info:PropTypes.object,
      }),
  };
  
  state={
    info:null,
    bigImg:0,
    content:0,
    amount:null,
    n:0,
  }
  
  componentDidMount(){
    voteEvents.addListener('AddBusket',this.addToBusket);
    let hash={};
    console.log(this.props.products.products);
    this.props.products.products.filter(v=>v.code==this.props.match.params.id)[0].stock.forEach(v => {
      hash[v.vol]=1;
    });
    this.setState({amount:hash})
  }
  componentWillUnmount = () => {
    voteEvents.removeListener('AddBusket',this.addToBusket);
  };

  
  
  setBigImg=(EO)=>{             
  this.setState({bigImg:EO.target.alt})
  }

  setContent=(EO)=>{
    this.setState({content:Number(EO.target.id)})
  }
  //  обработчик кнопки -
  dec=(EO)=>{   
    let newState={...this.state.amount};
    if(0<newState[parseInt(EO.target.id)]){
      newState[parseInt(EO.target.id)]-=1;
    }
    this.setState({amount:newState}); 
  }
 //  обработчик кнопки +
  inc=(EO)=>{
    let newState={...this.state.amount};
    let prod=this.props.products.products.filter(v=>v.code==Number(this.props.match.params.id))[0].stock.filter(v=>v.vol==parseInt(EO.target.id))[0];
    if(prod.in>newState[parseInt(EO.target.id)]){
      newState[parseInt(EO.target.id)]+=1;
    }
    this.setState({amount:newState}); 
  }
 
 
  setAmount=(EO)=>{
    let newState={...this.state.amount};
    newState[EO.target.id]=Number(EO.target.value);
    this.setState({amount:newState});
  }
  addToBusket=(EO)=>{
    if(EO.target.id>0){
    if (this.props.user.info!==null){
      let product=this.props.user.info.order;

      if ([this.props.match.params.id] in product ){
        if ([EO.target.id] in product[this.props.match.params.id]){
          product[this.props.match.params.id][EO.target.id]+=this.state.amount[EO.target.id];
        }
        else{
          product[this.props.match.params.id][EO.target.id]=this.state.amount[EO.target.id];
        }
      }
      else{
        product[this.props.match.params.id]={[EO.target.id]:this.state.amount[EO.target.id]};
      }
      
      this.props.dispatch(product_addtobusket(product));
      firebase.firestore().doc('Users/Users').set({
        [this.props.user.info.email]:{...this.props.user.info,order:product},
      })
    }
    else{
      if(localStorage.parfumShop_busket){

        let storage=JSON.parse(localStorage.parfumShop_busket);
        if ([this.props.match.params.id] in storage ){
          if ([EO.target.id] in storage[this.props.match.params.id]){
            storage[this.props.match.params.id][EO.target.id]+=this.state.amount[EO.target.id];
          }
          else{
            storage[this.props.match.params.id][EO.target.id]=this.state.amount[EO.target.id];
          }
        }
        else{
          storage[this.props.match.params.id]={[EO.target.id]:this.state.amount[EO.target.id]};
        }
 
        localStorage.setItem('parfumShop_busket',JSON.stringify(storage));
      }
      else{
        localStorage.setItem('parfumShop_busket',JSON.stringify({[this.props.match.params.id]:{[EO.target.id]:this.state.amount[EO.target.id]}}));
      }
      }
    }
  }

  render() {
    console.log('Page_Product render');

  //delete localStorage.parfumShop_busket;

    const id = Number(this.props.match.params.id);
    let product=this.props.products.products.filter(v=>v.code==id);

    return  <div className='PageProduct' >{product.map(v=>
      
      {
        let img=[];
        img.push(v.img.map((v,i)=><img alt={i} key={i} className='small' onClick={this.setBigImg} src={v}/>));
       
      return <div className='Product'>
        <Product amount={this.state.amount} img={img} prod={v}/>
        <div className="describe">
          <div className="Part">
            <span class={this.state.content==0?"title active":"title"} id="0" onClick={this.setContent}>Описание</span>
            <span class={this.state.content==1?"title active":"title"} id="1" onClick={this.setContent}>О бренде</span>
          </div>
          <div className="content">
          {
            this.state.content?
            <div>{this.props.brands.brands.filter(y=>y.name==v.brand).map(y=><p>{y.info}</p>)}</div>
            :
            <div><p key="name"><span class="bold">Полное наименование:</span> {v.name}</p>
            <p key="brand"><span class="bold">Бренд:</span> {v.brand}</p>
            <p key="info">{v.info }</p>
            </div>
          }
        </div>
      </div>
      {
        this.props.products.products.filter(v=>(product[0].brand==v.brand&&v.code!=product[0].code)).length!=0?
        <div class="ProdBrand">
         <h6>Другие товары фирмы {product[0].brand}</h6>
          <div>
          {
            this.props.products.products.filter(v=>(product[0].brand==v.brand&&v.code!=product[0].code)).map(v=>{
              return <div key={v.code} className='CatalogProduct'>
              <NavLink key={v.code} to={"/catalog/product/"+v.code} exact className="PageLink" activeClassName="ActivePageLink"><img alt={0} key={0}  src={v.img[0]}/>
              <h3>{v.brand}</h3>
              <p>{v.name}</p>
              <div class="Volume">Объем, мл: {v.stock.sort((a,b)=>a.vol-b.vol).map(v=><span key={v.vol}>{v.vol} </span>)}
              <div class="Var">{v.stock.length} варианта</div></div>
              <div class="Price">{Math.min.apply(null,v.stock.map(v=>v.price))} - {Math.max.apply(null,v.stock.map(v=>v.price))} руб.</div>
              <input type="button" value="Выбрать" class="hidden"/>
              </NavLink>
            </div>
          })
        }
        </div>
        </div>
        :
        null
      }
      
      </div>})}
  </div>;
   
    

  }

}

const mapStateToProps = function (state) {
  return {
    brands: state.brands,  // Магазины
    products: state.products,   //Товары
    user: state.user,
  };
};

export default connect(mapStateToProps)(Page_Product);
