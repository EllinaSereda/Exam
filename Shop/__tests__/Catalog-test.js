"use strict";

import React from 'react';
import renderer from 'react-test-renderer';
import Catalog from '../components/Catalog/Catalog';
import { BrowserRouter } from 'react-router-dom';

test('работа Main', () => {

  // создаём тестовую версию компонента
  const component = renderer.create(
    <BrowserRouter><Catalog products={[{brand:"Carol",code:0,img:["#","#"],info:"parfume",name:"Good",sex:"w",stock:[{in:0,price:122,vol:30}]}]}/></BrowserRouter>
  );

  // получаем снэпшот (HTML-снимок) компонента для сверки, что вёрстка не испортилась
  let componentTree=component.toJSON();
  expect(componentTree).toMatchSnapshot();

  // найдём в вёрстке компонента саму кнопку

  // получаем уже изменённый снэпшот


  
    
});
