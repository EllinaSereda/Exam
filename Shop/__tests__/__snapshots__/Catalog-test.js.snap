// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`работа Main 1`] = `
<div
  className="Catalog"
>
  <div
    className="Product"
  >
    <a
      aria-current={null}
      className="PageLink"
      href="/catalog/product/0"
      onClick={[Function]}
      style={undefined}
    >
      <img
        alt={0}
        src="#"
      />
      <h3>
        Carol
      </h3>
      <p>
        Good
      </p>
      <div
        class="Volume"
      >
        Объем, мл: 
        <span>
          30
           
        </span>
        <div
          class="Var"
        >
          1
           варианта
        </div>
      </div>
      <div
        class="Price"
      >
        122
         - 
        122
         руб.
      </div>
      <input
        class="hidden"
        type="button"
        value="Выбрать"
      />
    </a>
  </div>
</div>
`;
